pkg/controller/common.go

```diff
@@ -597,335 +600,805 @@ func (c *SharedState) isVolumeVisited(azVolumeName string) bool {
-			klog.V(5).Infof("Error listing AzVolumeAttachments for azvolume %s. Error: %v.", volume, err)
-		}
+func (p *podTolerationFilter) name() string {
+	return "pod toleration filter"
+}
 
-		klog.V(5).Infof("Found %d AzVolumeAttachments for volume %s. AzVolumeAttachments: %+v.", len(azVolumeAttachments), volume, azVolumeAttachments)
-		for _, azVolumeAttachment := range azVolumeAttachments {
-			if azVolumeAttachment.Spec.RequestedRole == diskv1alpha2.PrimaryRole {
-				klog.V(5).Infof("AzVolumeAttachments %s for volume %s is primary on node %s.", azVolumeAttachment.Name, volume, azVolumeAttachment.Spec.NodeName)
-				if primaryNode == "" {
-					primaryNode = azVolumeAttachment.Spec.NodeName
-				}
+func (p *podTolerationFilter) setup(pods []*v1.Pod, persistentVolumes []*v1.PersistentVolume, nodes []v1.Node, azr azReconciler) {
+	p.pods = pods
+}
+
+func (p *podTolerationFilter) filter(ctx context.Context, nodes []v1.Node) ([]v1.Node, error) {
+	candidateNodes := set{}
+	for i := range nodes {
+		candidateNodes.add(i)
+	}
+
+	podTolerations := set{}
+
+	for i, pod := range p.pods {
+		podTolerationMap := map[string]*v1.Toleration{}
+		for _, podToleration := range pod.Spec.Tolerations {
+			podToleration := &podToleration
+			if i == 0 {
+				podTolerations.add(podToleration)
 			} else {
-				klog.V(5).Infof("Azvolumeattachment %s is not primary. Scoring node %s with replica attachment. Current score array is %v. Len: %d.", azVolumeAttachment.Name, azVolumeAttachment.Spec.NodeName, nodeScores, len(nodeScores))
-				if _, ok := nodeScores[azVolumeAttachment.Spec.NodeName]; !ok {
-					// Node is not in the array. Adding node to the array with score 0.
-					nodeScores[azVolumeAttachment.Spec.NodeName] = 0
-					nodes = append(nodes, azVolumeAttachment.Spec.NodeName)
+				podTolerationMap[podToleration.Key] = podToleration
+			}
+		}
+		if i > 0 {
+			for podToleration := range podTolerations {
+				if existingToleration, ok := podTolerationMap[podToleration.(v1.Toleration).Key]; ok {
+					if !podToleration.(*v1.Toleration).MatchToleration(existingToleration) {
+						podTolerations.remove(podToleration)
+					}
 				}
-				nodeScores[azVolumeAttachment.Spec.NodeName]++
-				klog.V(5).Infof("New score for node %s is %d. Updated score array is %v. Len: %d.", azVolumeAttachment.Spec.NodeName, nodeScores[azVolumeAttachment.Spec.NodeName], nodeScores, len(nodeScores))
 			}
 		}
+	}
 
-		var pv v1.PersistentVolume
-		if err := azr.getClient().Get(ctx, types.NamespacedName{Name: azVolume.Status.PersistentVolume}, &pv); err != nil {
-			return nil, err
+	for candidateNode := range candidateNodes {
+		tolerable := true
+		node := nodes[candidateNode.(int)]
+		for _, taint := range node.Spec.Taints {
+			taintTolerable := false
+			for podToleration := range podTolerations {
+				// if any one of node's taint cannot be tolerated by pod's tolerations, break
+				if podToleration.(*v1.Toleration).ToleratesTaint(&taint) {
+					taintTolerable = true
```

Comment:
Should we set taintTolerable=true at the beginning of the loop and if !podToleration(*v1.Toleration).ToleratesTaint(&taint) and then then set taintTolerable=false here and break. I suppose even if one of the podToleration can't tolerate a taint, we should break?

Reply:
if `taintTolerable` is set `true` at the beginning of the loop, we would end up with `tolerable == true` when `len(podTolerations) == 0`. Good catch on the `break`, I don't know where that wen haha.